File: app/src/main/java/eu/napcode/gonoteit/ui/login/LoginViewModel.java
package eu.napcode.gonoteit.ui.login;

import android.annotation.SuppressLint;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import javax.inject.Inject;

import eu.napcode.gonoteit.AuthenticateMutation;
import eu.napcode.gonoteit.repository.Resource;
import eu.napcode.gonoteit.repository.user.UserRepository;
import eu.napcode.gonoteit.rx.RxSchedulers;
import timber.log.Timber;

public class LoginViewModel extends ViewModel {

    private UserRepository userRepository;
    private UserValidator userValidator;
    private RxSchedulers rxSchedulers;

    private String login, password;

    private MutableLiveData<Boolean> inputsValid = new MutableLiveData<>();
    private MutableLiveData<Resource<AuthenticateMutation.Data>> authenticationResource = new MutableLiveData<>();

    @Inject
    public LoginViewModel(UserRepository userRepository, UserValidator userValidator, RxSchedulers rxSchedulers) {
        this.userRepository = userRepository;
        this.userValidator = userValidator;
        this.rxSchedulers = rxSchedulers;

        init();
    }

    private void init() {
        validate();
    }

    public LiveData<Boolean> areInputsValid() {
        return inputsValid;
    }

    public void setLogin(String login) {
        this.login = login;
        validate();
    }

    private void validate() {
        boolean valid = userValidator.isUserValid(login, password);
        Timber.d("Validated input %s and %s. Result: %b", login, password, valid);

        inputsValid.postValue(valid);
    }

    public void setPassword(String password) {
        this.password = password;
        validate();
    }

    @SuppressLint("CheckResult")
    public MutableLiveData<Resource<AuthenticateMutation.Data>> login() {
        userRepository
                .authenticateUser(login, password, authenticationResource)
                .subscribeOn(rxSchedulers.io())
                .observeOn(rxSchedulers.androidMainThread())
                .doOnSubscribe(it -> authenticationResource.postValue(Resource.loading(null)))
                .doOnError(error -> {
                    authenticationResource.postValue(Resource.error(error));
                    Timber.d("Login error: %s", error.getLocalizedMessage());
                })
                .subscribe();

        return authenticationResource;
    }

}

File: app/src/main/java/eu/napcode/gonoteit/ui/login/LoginViewModel.kt
package eu.napcode.gonoteit.ui.login

import android.annotation.SuppressLint
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel

import javax.inject.Inject

import eu.napcode.gonoteit.AuthenticateMutation
import eu.napcode.gonoteit.app.utils.isHostValid
import eu.napcode.gonoteit.app.utils.isUserValid
import eu.napcode.gonoteit.repository.Resource
import eu.napcode.gonoteit.repository.user.UserRepository
import eu.napcode.gonoteit.rx.RxSchedulers
import timber.log.Timber

class LoginViewModel @Inject
constructor(private val userRepository: UserRepository, private val rxSchedulers: RxSchedulers) : ViewModel() {

    private var login: String? = null
    private var password: String? = null
    private var host: String? = null

    private val inputsValid = MutableLiveData<Boolean>()
    private val authenticationResource = MutableLiveData<Resource<AuthenticateMutation.Data>>()

    init {
        validate()
    }

    fun areInputsValid(): LiveData<Boolean> {
        return inputsValid
    }

    fun setLogin(login: String) {
        this.login = login
        validate()
    }

    fun setHost(host: String) {
        this.host = host
        validate()
    }

    private fun validate() {
        val valid = isUserValid(login, password) && isHostValid(host)
        Timber.d("Validated input %s and %s. Result: %b", login, password, valid)

        inputsValid.postValue(valid)
    }

    fun setPassword(password: String) {
        this.password = password
        validate()
    }

    @SuppressLint("CheckResult")
    fun login(): MutableLiveData<Resource<AuthenticateMutation.Data>> {
        userRepository
                .authenticateUser(host!!, login!!, password!!, authenticationResource)
                .subscribeOn(rxSchedulers.io())
                .observeOn(rxSchedulers.androidMainThread())
                .doOnSubscribe { authenticationResource.postValue(Resource.loading<AuthenticateMutation.Data>(null)) }
                .doOnError { error ->
                    authenticationResource.postValue(Resource.error<AuthenticateMutation.Data>(error))
                    Timber.d("Login error: %s", error.localizedMessage)
                }
                .subscribe()

        return authenticationResource
    }

}